packages/flutter_tools/gradle/flutter.gradle

```diff
@@ -817,8 +840,56 @@ class FlutterPlugin implements Plugin<Project> {
                     }
                 }
             }
-        } else {
-            project.android.libraryVariants.all addFlutterDeps
+            configurePlugins()
+            return
+        }
+        // Flutter host module project (Add-to-app).
+        String hostAppProjectName = project.rootProject.hasProperty('flutter.hostAppProjectName') ? project.rootProject.property('flutter.hostAppProjectName') : "app"
+        Project appProject = project.rootProject.findProject(":${hostAppProjectName}")
+        assert appProject != null : "Project :${hostAppProjectName} doesn't exist. To custom the host app project name, set `org.gradle.project.flutter.hostAppProjectName=<project-name>` in gradle.properties."
+        // Wait for the host app project configuration.
+        appProject.afterEvaluate {
+            assert appProject.android != null
+            project.android.libraryVariants.all { libraryVariant ->
+                Task copyFlutterAssetsTask
+                appProject.android.applicationVariants.all { appProjectVariant ->
+                    Task appAssembleTask = getAssembleTask(appProjectVariant)
+                    if (!shouldConfigureFlutterTask(appAssembleTask)) {
+                        return
+                    }
+                    // Find a compatible application variant in the host app.
+                    //
+                    // For example, consider a host app that defines the following variants:
+                    // | ----------------- | ----------------------------- |
+                    // |   Build Variant   |   Flutter Equivalent Variant  |
+                    // | ----------------- | ----------------------------- |
+                    // |   freeRelease     |   relese                      |
+                    // |   freeDebug       |   debug                       |
+                    // |   freeDevelop     |   debug                       |
+                    // |   profile         |   profile                     |
+                    // | ----------------- | ----------------------------- |
+                    //
+                    // This mapping is based on the following rules:
+                    // 1. If the host app build variant name is `profile` then the equivalent
+                    //    Flutter variant is `profile`.
+                    // 2. If the host app build variant is debuggable
+                    //    (e.g. `buildType.debuggable = true`), then the equivalent Flutter
+                    //    variant is `debug`.
+                    // 3. Otherwise, the equivalent Flutter variant is `release`.
+                    String variantBuildMode = buildModeFor(libraryVariant.buildType)
+                    if (buildModeFor(appProjectVariant.buildType) != variantBuildMode) {
+                        return
```

Comment:
Can we print a message here? I always found the mismatch error a bit cryptic.

Reply:
This is during configuration phase. It could be quite spammy, and no very meaningful for debugging. What's the user story?

Reply:
Maybe I'm thinking of something else. In Android Studio, open build variants tab in case of gradle subproject integration, pick one  variant for the app and another variant for the flutter module.